HTMLScriptElement.src - update for TrustedScriptURL

hamishwillee opened the pull request:
This updates the TrustedScriptURL documentation for [`HTMLScriptElement/src`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLScriptElement/src) to include information about using it with trusted types.

I don't know why/when it would ever make sense to accept an  untrusted URL, and if you did, to use trusted types. I've presented this primarily as "Use CSP to restrict the script options but if you have to take and untrusted script then you can use TT".

Project tracking in #37518

Also part of #41507

hamishwillee committed:
HTMLScriptElement.src - update for TrustedScriptURL

hamishwillee commented on files/en-us/web/api/htmlscriptelement/src/index.md:
```diff
@@ -8,16 +8,102 @@ browser-compat: api.HTMLScriptElement.src
+### Exceptions
+
+- `TypeError`
+  - : Thrown if the property is set with a string when [Trusted Types](/en-US/docs/Web/API/Trusted_Types_API) are [enforced by a CSP](/en-US/docs/Web/API/Trusted_Types_API#using_a_csp_to_enforce_trusted_types) and no default policy is defined.
```
```suggestion
  - : Thrown if the property is set with a string when [Trusted Types](/en-US/docs/Web/API/Trusted_Types_API) are [enforced by a CSP](/en-US/docs/Web/API/Trusted_Types_API#using_a_csp_to_enforce_trusted_types) and no default policy is defined.
    This is also thrown if the fetched URL cannot be successfully parsed as its indicated type, such as a module or importmap.
  
```

hamishwillee commented on files/en-us/web/api/htmlscriptelement/src/index.md:
```diff
@@ -8,16 +8,104 @@ browser-compat: api.HTMLScriptElement.src
+
+- `TypeError`
+  - : Thrown if the property is set with a string when [Trusted Types](/en-US/docs/Web/API/Trusted_Types_API) are [enforced by a CSP](/en-US/docs/Web/API/Trusted_Types_API#using_a_csp_to_enforce_trusted_types) and no default policy is defined.
+    This is also thrown if the fetched URL cannot be successfully parsed as its indicated type, such as a module or importmap.
```
This isn't quite right, but I'm not sure how to be precise.
I believe it is a NetworkError on fetching. TypeError is thrown for an importmap during parsing (https://html.spec.whatwg.org/multipage/webappapis.html#import-map-authoring-requirements), and I think it is a Syntax error you get normally on parsing JavaScript.

Thoughts?

hamishwillee replied:
Sorry. I think the text is correct. There are however other errors you might get such as NetworkError and SyntaxError.
I am not going to capture those here on the grounds that this fix is "better than it was" and I have lots to do in this release cycle. Resolving.

wbamberg commented on files/en-us/web/api/htmlscriptelement/src/index.md:
```diff
@@ -8,16 +8,103 @@ browser-compat: api.HTMLScriptElement.src
 
+> [!WARNING]
+> This property represents the URI of an external script loaded into the script element, which may be executable depending on the script {{domxref("HTMLScriptElement/type","type")}}.
+> APIs like this are known as [injection sinks](/en-US/docs/Web/API/Trusted_Types_API#concepts_and_usage), and are potentially a vector for [cross-site-scripting (XSS)](/en-US/docs/Web/Security/Attacks/XSS) and other types of [website attacks](/en-US/docs/Web/Security/Attacks).
```
I should have noticed this before but it's not "cross-site-scripting", it's "cross-site scripting".

Also, which other types of website attacks?

wbamberg commented on files/en-us/web/api/htmlscriptelement/src/index.md:
```diff
@@ -8,16 +8,103 @@ browser-compat: api.HTMLScriptElement.src
+
+The `src` property is used to load and run external scripts.
+The fetched script is run in the context of the current page, and can hence do anything that your own website code can do (even if the URL is not same-origin with your site).
+If the input is provided by a user, this is a possible vector for [Cross-site-scripting (XSS)](/en-US/docs/Web/Security/Attacks/XSS) attacks.
```
```suggestion
If the input is provided by a user, this is a possible vector for [Cross-site scripting (XSS)](/en-US/docs/Web/Security/Attacks/XSS) attacks.
```

wbamberg commented on files/en-us/web/api/htmlscriptelement/src/index.md:
```diff
@@ -8,16 +8,103 @@ browser-compat: api.HTMLScriptElement.src
+const trustedScriptURL = policy.createScriptURL(untrustedScript);
+
+// Inject the TrustedScriptURL (which contains a trusted URL)
+el.src = trustedScriptURL;
```
I mean it seems unfortunate that this TT is going to allow that naughty JS.

hamishwillee replied:
What do you mean? `trustedScriptURL` is either an URL we are OK with in scriptAllowList  or the empty string.

wbamberg replied:
Yes, I think this was from before you updated the policy to have an allowlist!

hamishwillee committed:
Fix all typo cross-site-scripting

hamishwillee committed:
Update files/en-us/web/api/htmlscriptelement/src/index.md

Co-authored-by: wbamberg <will@bootbonnet.ca>

hamishwillee commented on files/en-us/web/api/htmlscriptelement/src/index.md:
```diff
@@ -8,16 +8,103 @@ browser-compat: api.HTMLScriptElement.src
+If you find a situation where this is necessary, you should always assign {{domxref("TrustedScriptURL")}} objects instead of strings, and [enforce trusted type](/en-US/docs/Web/API/Trusted_Types_API#using_a_csp_to_enforce_trusted_types) using the [`require-trusted-types-for`](/en-US/docs/Web/HTTP/Reference/Headers/Content-Security-Policy/require-trusted-types-for) CSP directive.
+This ensures that the input is passed through a transformation function, which has the chance to [sanitize](/en-US/docs/Web/Security/Attacks/XSS#sanitization) or reject the URL before it is injected.
+
+The behavior of the transformation function will depend on the specific use case that requires a user provided script.
```
@wbamberg The problem here is that I don't see any "real" reason why you would **ever** use trusted types. If you want to allow only specified scripts to execute then you would use  CSP that defines a `script-src` - that would be more effective/less error prone that checking in your transformation function.
If you want to turn it off, same thing - set the CSP to none. 

The only "somewhat real" reason I can think of is that want to allow some scripts to run but not all the time. So you perhaps use TT to decide at a particular state of your code you'll allow an input.

Hence the "where it is necessary" words. 

So you can see I don't want to say "nowhere in my code do I want that to happen" because I could block that more effectively with `script-src`. This only kind of makes sense if you can't know what the URLs are going to be - and hence can't use a normal CSP with script-src.

Does that make sense?

wbamberg replied:
> @wbamberg
> 
> I removed "sanitized" and replaced with "This ensures that the input is passed through a transformation function, which has the chance to reject or modify the URL before it is injected.". The _modify_ is to catch cases where you might want to allow but redirect the URL or perhaps strip off URL parameters or whatever - I don't really have a practical case.
> 
> > For JavaScript and JavaScript URL sinks, the policy may turn off the sinks entirely or allow certain predefined inputs (for example, specific URLs).
> 
> The problem here is that if you want to block scripts or allow only specified scripts to execute then you would use CSP that defines a `script-src` - much is more effective and less error prone than using a transformation.
> 

I would like to know what @aaronshim thinks about this. Because we recommend that developers use a strict CSP, that uses nonces or hashes (depending on whether you can generate responses dynamically or not) rather than locations. But if you use nonces, then setting `src` is dangerous, because the script will be allowed to run if it also has the correct nonce (which it probably will, if the `<script>` element we are modifying was properly served in the first place). So at least in that case I could see a use for a TT policy that bans assignment to `src`.

I mean, as a developer, say you don't ever want to see assignment to `script.src`. Is that alone a strong enough reason to use a location-based CSP (with all its known fragility) instead of a nonce-based one?

> I don't see any "real" reason why you would **ever** use trusted types, other than you're already enforcing them for other things so you need to (perhaps I should say that). Hence the "where it is necessary" words.
> 
> Does that make sense?

hamishwillee replied:
Ah, yes I meant matching nonces (updated), yes that sounds like a risk, thanks, an interesting use case.

wbamberg replied:
FWIW Aaron had a related useful comment at https://github.com/mdn/content/pull/37917#discussion_r1958668942.

hamishwillee committed:
Security considerations update

hamishwillee commented on files/en-us/web/api/htmlscriptelement/src/index.md:
```diff
@@ -8,16 +8,103 @@ browser-compat: api.HTMLScriptElement.src
+The `src` property is used to load and run external scripts.
+The fetched script is run in the context of the current page, and can hence do anything that your own website code can do (even if the URL is not same-origin with your site).
+If the input is provided by a user, this is a possible vector for [Cross-site-scripting (XSS)](/en-US/docs/Web/Security/Attacks/XSS) attacks.
+Even if the resource is trusted by your website, it may still be compromised in a supply chain attack.
```
Yes, removed the nonce stuff and moved the supply chain stuff to the end. It is just links - no deep flavour.

hamishwillee committed:
Fix up example

wbamberg commented on files/en-us/web/api/htmlscriptelement/src/index.md:
```diff
@@ -8,16 +8,113 @@ browser-compat: api.HTMLScriptElement.src
+
+If you're using this property and [enforcing trusted types](/en-US/docs/Web/API/Trusted_Types_API#using_a_csp_to_enforce_trusted_types) (using the [`require-trusted-types-for`](/en-US/docs/Web/HTTP/Reference/Headers/Content-Security-Policy/require-trusted-types-for) CSP directive), you will need to always assign {{domxref("TrustedScriptURL")}} objects instead of strings.
+This ensures that the input is passed through a transformation function, which has the chance to reject or modify the URL before it is injected.
+If you're permitting some sources using `CSP: script-src`, the transformation function would use the sources as an allow-list.
```
```suggestion
```

I would omit this, because not necessarily: you might be happy to allow some scripts in the markup you serve but not want to allow their dynamic assignment.

wbamberg commented on files/en-us/web/api/htmlscriptelement/src/index.md:
```diff
@@ -8,16 +8,113 @@ browser-compat: api.HTMLScriptElement.src
+Next we create a {{domxref("TrustedTypePolicy")}} that defines a {{domxref("TrustedTypePolicy/createScriptURL", "createScriptURL()")}} method for transforming input strings into {{domxref("TrustedScriptURL")}} instances.
+
+For the purpose of this example we'll assume that we want to allow a set URLs in the `scriptAllowList` array and log any other scripts.
+Note that if we were using a `CSP: script-src` we might just be able to pass through all URLs.
```
```suggestion
```

...this is not good advice AFAICT. If we have `script-src` with a nonce, which is our recommendation, then it is surely dangerous to pass through all URLs.

wbamberg reviewed (commented):
I made a couple of suggestions to downplay the stuff about CSP here.

hamishwillee committed:
Apply suggestions from code review

Co-authored-by: wbamberg <will@bootbonnet.ca>

hamishwillee committed:
Clarify example for reading the `src` property

Updated the section on reading the `src` property to clarify the example and its context.

wbamberg reviewed (approved):
👍 thanks Hamish.
